x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteStep.java

```diff
@@ -40,11 +41,21 @@ public void performDuringNoSnapshot(IndexMetadata indexMetadata, ClusterState cu
         DataStream dataStream = indexAbstraction.getParentDataStream();
 
         if (dataStream != null) {
-            assert dataStream.getWriteIndex() != null : dataStream.getName() + " has no write index";
+            final Index targetWriteIndex;
+            final boolean isFailureStoreIndex;
+            if (dataStream.isFailureStoreIndex(indexName)) {
+                targetWriteIndex = dataStream.getFailureStoreWriteIndex();
+                isFailureStoreIndex = true;
+            } else {
+                targetWriteIndex = dataStream.getWriteIndex();
+                isFailureStoreIndex = false;
+            }
+            assert targetWriteIndex != null : dataStream.getName() + " has no write index";
 
             // using index name equality across this if/else branch as the UUID of the index might change via restoring a data stream
             // with one index from snapshot
-            if (dataStream.getIndices().size() == 1 && dataStream.getWriteIndex().getName().equals(indexName)) {
+            int totalDataStreamIndices = dataStream.getIndices().size() + dataStream.getFailureIndices().getIndices().size();
+            if (totalDataStreamIndices == 1 && targetWriteIndex.getName().equals(indexName)) {
                 // This is the last index in the data stream, the entire stream
                 // needs to be deleted, because we can't have an empty data stream
```

Comment:
I think this might cause some issues in the following scenario: we have 1 backing index and 1 failure index, and ILM wants to delete the backing index, but will end up in an error state due to the next `else if` branch. This means data might not get deleted when users expect it to.

At the same time, I'm not sure simply ignoring the failure store indices and just deleting the whole data stream is a good alternative either, as users might still want to have those failures...

I'll think some more about possible options, but in the meantime, I'll put this up for discussion already.

Reply:
Ha, I didn't think this was possible. If I am not mistaken, the only way to get in this situation is if there is a policy that only has a delete phase and nothing else. Anything else would require a rollover to proceed, right @jbaiera ?

If my understanding is correct I see the following 2 scenarios on a single data stream:
- ILM tries to delete the write backing index, in this case I think we should delete the whole data stream and ignore the failure store. Consider that there is no guarantee that this data stream is empty, the write index might have data, but the user for whatever reason wanted it deleted, so if they do not care about their "happy" data why would they care about the failure?
- ILM tries to delete the write failure index, in this case I think we should fail and wait for the backing index to be deleted and trigger the data stream deletion. At least I think this is our way forward until we implement defaulting to no failure indices.

What do you guys think?

Reply:
I think I put check together really quickly without considering the wider cases. I agree that we should just delete the entire data stream if there's a singular write backing index and simply fail if we try to delete the failure write index. We can circle back on this for clearing out old failure indices later.

Reply:
Hmm, I was adding additional commentary in the code to explain the logic a bit more, and wrote out `We [delete the entire data stream] even if there are multiple failure store indices because otherwise we would never delete the index.`

We don't consider this a potential data loss problem in regards to the failure store do we? On one hand, why wouldn't you want to keep those records around in case you need to remediate them. On the other hand, if those writes were successful, they'd be deleted all the same when the delete step happens. I'm going to air on the side of moving forward with the suggested change, but wanted to bring this up to make sure we're aligned on the thinking here.

Reply:
> On the other hand, if those writes were successful, they'd be deleted all the same when the delete step happens

This is exactly what I was thinking.